File: saleads-e2e/scripts/saleads_mi_negocio_full_test.mjs
#!/usr/bin/env node

import fs from "node:fs/promises";
import path from "node:path";
import process from "node:process";
import { chromium } from "playwright";

const REPORT_FIELDS = [
  "Login",
  "Mi Negocio menu",
  "Agregar Negocio modal",
  "Administrar Negocios view",
  "Información General",
  "Detalles de la Cuenta",
  "Tus Negocios",
  "Términos y Condiciones",
  "Política de Privacidad"
];

const GOOGLE_ACCOUNT_EMAIL =
  process.env.SALEADS_GOOGLE_ACCOUNT ?? "juanlucasbarbiergarzon@gmail.com";

function parseArgs() {
  const args = process.argv.slice(2);
  const parsed = {
    url: process.env.SALEADS_LOGIN_URL ?? "",
    headless: process.env.HEADLESS === "true",
    slowMo: Number(process.env.SALEADS_SLOW_MO ?? "120")
  };

  for (let i = 0; i < args.length; i += 1) {
    const arg = args[i];
    if (arg === "--url" && args[i + 1]) {
      parsed.url = args[i + 1];
      i += 1;
      continue;
    }

    if (arg === "--headless") {
      parsed.headless = true;
      continue;
    }

    if (arg === "--headed") {
      parsed.headless = false;
      continue;
    }
  }

  return parsed;
}

function nowTag() {
  return new Date().toISOString().replaceAll(":", "-").replaceAll(".", "-");
}

function createDefaultReport() {
  return REPORT_FIELDS.reduce((acc, field) => {
    acc[field] = {
      status: "FAIL",
      details: "Not executed."
    };
    return acc;
  }, {});
}

async function ensureDirectory(dirPath) {
  await fs.mkdir(dirPath, { recursive: true });
}

async function waitForUi(page) {
  await page.waitForTimeout(500);
  await page.waitForLoadState("domcontentloaded", { timeout: 15000 }).catch(
    () => {}
  );
  await page.waitForLoadState("networkidle", { timeout: 8000 }).catch(() => {});
}

async function saveScreenshot(page, evidenceDir, fileName, fullPage = false) {
  const outputPath = path.join(evidenceDir, `${fileName}.png`);
  await page.screenshot({ path: outputPath, fullPage });
  return outputPath;
}

async function pickVisible(page, description, candidates, timeoutMs = 20000) {
  const deadline = Date.now() + timeoutMs;

  while (Date.now() < deadline) {
    for (const candidate of candidates) {
      const first = candidate.first();
      const count = await first.count();
      if (!count) {
        continue;
      }

      const visible = await first.isVisible().catch(() => false);
      if (visible) {
        return first;
      }
    }

    await page.waitForTimeout(250);
  }

  throw new Error(`Timed out waiting for visible element: ${description}`);
}

async function clickAndWait(page, locator) {
  await locator.click();
  await waitForUi(page);
}

async function assertVisible(page, description, candidates, timeoutMs = 20000) {
  return pickVisible(page, description, candidates, timeoutMs);
}

async function clickGoogleLogin(mainPage, context) {
  const googleButton = await assertVisible(
    mainPage,
    "Google login button",
    [
      mainPage.getByRole("button", {
        name: /sign in with google|iniciar sesión con google|continuar con google/i
      }),
      mainPage.getByRole("link", {
        name: /sign in with google|iniciar sesión con google|continuar con google/i
      }),
      mainPage.getByRole("button", { name: /google/i }),
      mainPage.getByText(
        /sign in with google|iniciar sesión con google|continuar con google/i
      )
    ],
    25000
  );

  const popupPromise = context
    .waitForEvent("page", { timeout: 8000 })
    .catch(() => null);

  await clickAndWait(mainPage, googleButton);

  const popupPage = await popupPromise;
  return popupPage;
}

async function maybePickGoogleAccount(page) {
  await waitForUi(page);
  const emailMatcher = new RegExp(
    GOOGLE_ACCOUNT_EMAIL.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"),
    "i"
  );

  const accountCandidate = await pickVisible(
    page,
    "Google account selector",
    [
      page.getByRole("button", { name: emailMatcher }),
      page.getByText(emailMatcher),
      page.getByRole("link", { name: emailMatcher })
    ],
    7000
  ).catch(() => null);

  if (!accountCandidate) {
    return false;
  }

  await clickAndWait(page, accountCandidate);
  return true;
}

async function waitForMainApp(page) {
  const sidebar = await assertVisible(
    page,
    "left sidebar navigation",
    [
      page.locator("aside"),
      page.getByRole("navigation"),
      page.locator('[class*="sidebar"]')
    ],
    70000
  );

  const negocioEntry = await assertVisible(
    page,
    "Negocio navigation entry",
    [page.getByText(/^negocio$/i), page.getByRole("button", { name: /^negocio$/i })],
    20000
  );

  return { sidebar, negocioEntry };
}

async function openMiNegocioMenu(page) {
  const negocioEntry = await assertVisible(
    page,
    "Negocio entry",
    [page.getByRole("button", { name: /^negocio$/i }), page.getByText(/^negocio$/i)]
  );
  await clickAndWait(page, negocioEntry);

  const miNegocioEntry = await assertVisible(
    page,
    "Mi Negocio entry",
    [
      page.getByRole("button", { name: /^mi negocio$/i }),
      page.getByRole("link", { name: /^mi negocio$/i }),
      page.getByText(/^mi negocio$/i)
    ]
  );
  await clickAndWait(page, miNegocioEntry);

  const agregar = await assertVisible(
    page,
    "Agregar Negocio submenu item",
    [
      page.getByRole("button", { name: /^agregar negocio$/i }),
      page.getByRole("link", { name: /^agregar negocio$/i }),
      page.getByText(/^agregar negocio$/i)
    ]
  );

  const administrar = await assertVisible(
    page,
    "Administrar Negocios submenu item",
    [
      page.getByRole("button", { name: /^administrar negocios$/i }),
      page.getByRole("link", { name: /^administrar negocios$/i }),
      page.getByText(/^administrar negocios$/i)
    ]
  );

  return { agregar, administrar };
}

async function openAgregarNegocioModal(page) {
  const agregarNegocio = await assertVisible(
    page,
    "Agregar Negocio option",
    [
      page.getByRole("button", { name: /^agregar negocio$/i }),
      page.getByRole("link", { name: /^agregar negocio$/i }),
      page.getByText(/^agregar negocio$/i)
    ]
  );
  await clickAndWait(page, agregarNegocio);

  await assertVisible(
    page,
    "Crear Nuevo Negocio modal title",
    [page.getByRole("heading", { name: /^crear nuevo negocio$/i }), page.getByText(/^crear nuevo negocio$/i)]
  );
  const nombreInput = await assertVisible(
    page,
    "Nombre del Negocio input",
    [
      page.getByLabel(/nombre del negocio/i),
      page.getByPlaceholder(/nombre del negocio/i),
      page.locator("input").filter({ hasText: /nombre del negocio/i })
    ]
  ).catch(async () => {
    const labeledInput = page
      .locator("label")
      .filter({ hasText: /nombre del negocio/i })
      .locator("..")
      .locator("input");
    return assertVisible(page, "Nombre del Negocio input (fallback)", [labeledInput]);
  });

  await assertVisible(page, "usage text 2 de 3 negocios", [
    page.getByText(/tienes\s*2\s*de\s*3\s*negocios/i)
  ]);
  const cancelar = await assertVisible(page, "Cancelar button", [
    page.getByRole("button", { name: /^cancelar$/i }),
    page.getByText(/^cancelar$/i)
  ]);
  await assertVisible(page, "Crear Negocio button", [
    page.getByRole("button", { name: /^crear negocio$/i }),
    page.getByText(/^crear negocio$/i)
  ]);

  await nombreInput.fill("Negocio Prueba Automatización");
  await waitForUi(page);

  return { cancelar };
}

async function openAdministrarNegocios(page) {
  const miNegocio = await assertVisible(
    page,
    "Mi Negocio entry",
    [
      page.getByRole("button", { name: /^mi negocio$/i }),
      page.getByRole("link", { name: /^mi negocio$/i }),
      page.getByText(/^mi negocio$/i)
    ]
  );
  await clickAndWait(page, miNegocio);

  const administrar = await assertVisible(
    page,
    "Administrar Negocios option",
    [
      page.getByRole("button", { name: /^administrar negocios$/i }),
      page.getByRole("link", { name: /^administrar negocios$/i }),
      page.getByText(/^administrar negocios$/i)
    ]
  );
  await clickAndWait(page, administrar);

  await assertVisible(page, "Información General section", [
    page.getByRole("heading", { name: /información general/i }),
    page.getByText(/información general/i)
  ]);
  await assertVisible(page, "Detalles de la Cuenta section", [
    page.getByRole("heading", { name: /detalles de la cuenta/i }),
    page.getByText(/detalles de la cuenta/i)
  ]);
  await assertVisible(page, "Tus Negocios section", [
    page.getByRole("heading", { name: /tus negocios/i }),
    page.getByText(/tus negocios/i)
  ]);
  await assertVisible(page, "Sección Legal section", [
    page.getByRole("heading", { name: /sección legal/i }),
    page.getByText(/sección legal/i)
  ]);
}

async function validateInformacionGeneral(page) {
  await assertVisible(page, "user name", [
    page.locator('[class*="name"]'),
    page.getByText(/@/).locator("..")
  ]);
  await assertVisible(page, "user email", [
    page.getByText(/@/),
    page.locator("a[href^='mailto:']")
  ]);
  await assertVisible(page, "BUSINESS PLAN text", [
    page.getByText(/business plan/i)
  ]);
  await assertVisible(page, "Cambiar Plan button", [
    page.getByRole("button", { name: /cambiar plan/i }),
    page.getByText(/cambiar plan/i)
  ]);
}

async function validateDetallesCuenta(page) {
  await assertVisible(page, "Cuenta creada text", [page.getByText(/cuenta creada/i)]);
  await assertVisible(page, "Estado activo text", [
    page.getByText(/estado activo/i),
    page.getByText(/activo/i)
  ]);
  await assertVisible(page, "Idioma seleccionado text", [
    page.getByText(/idioma seleccionado/i)
  ]);
}

async function validateTusNegocios(page) {
  await assertVisible(page, "business list", [
    page.locator("table"),
    page.locator("ul").filter({ hasText: /negocio/i }),
    page.getByText(/tus negocios/i).locator("..")
  ]);
  await assertVisible(page, "Agregar Negocio button", [
    page.getByRole("button", { name: /^agregar negocio$/i }),
    page.getByText(/^agregar negocio$/i)
  ]);
  await assertVisible(page, "Tienes 2 de 3 negocios text", [
    page.getByText(/tienes\s*2\s*de\s*3\s*negocios/i)
  ]);
}

async function validateLegalLink({
  page,
  context,
  linkTextRegex,
  headingRegex,
  evidenceDir,
  screenshotName
}) {
  const appPageUrlBefore = page.url();
  const legalLink = await assertVisible(page, "legal link", [
    page.getByRole("link", { name: linkTextRegex }),
    page.getByText(linkTextRegex)
  ]);

  const popupPromise = context
    .waitForEvent("page", { timeout: 7000 })
    .catch(() => null);
  await clickAndWait(page, legalLink);
  const popupPage = await popupPromise;

  const legalPage = popupPage ?? page;
  await waitForUi(legalPage);

  await assertVisible(legalPage, "legal heading", [
    legalPage.getByRole("heading", { name: headingRegex }),
    legalPage.getByText(headingRegex)
  ]);

  const bodyText = (await legalPage.locator("body").innerText()).trim();
  if (bodyText.length < 150) {
    throw new Error("Legal page content is too short.");
  }

  const screenshotPath = await saveScreenshot(
    legalPage,
    evidenceDir,
    screenshotName,
    true
  );
  const finalUrl = legalPage.url();

  if (popupPage) {
    await page.bringToFront();
    await waitForUi(page);
  } else if (page.url() !== appPageUrlBefore) {
    await page.goBack().catch(() => null);
    await waitForUi(page);
  }

  return { screenshotPath, finalUrl };
}

function writeReportLine(report) {
  console.log("");
  console.log("========== SALEADS MI NEGOCIO FINAL REPORT ==========");
  for (const field of REPORT_FIELDS) {
    const entry = report[field];
    console.log(
      `${field}: ${entry.status}${entry.details ? ` - ${entry.details}` : ""}`
    );
  }
  console.log("=====================================================");
  console.log("");
}

async function run() {
  const options = parseArgs();
  const report = createDefaultReport();
  const runFolder = path.join(
    process.cwd(),
    "artifacts",
    "saleads_mi_negocio_full_test",
    nowTag()
  );
  await ensureDirectory(runFolder);

  const browser = await chromium.launch({
    headless: options.headless,
    slowMo: options.slowMo
  });
  const context = await browser.newContext({
    viewport: { width: 1440, height: 900 }
  });
  const page = await context.newPage();

  try {
    if (options.url) {
      await page.goto(options.url, { waitUntil: "domcontentloaded" });
      await waitForUi(page);
    }

    const popupPage = await clickGoogleLogin(page, context);
    if (popupPage) {
      await popupPage.waitForLoadState("domcontentloaded", { timeout: 20000 }).catch(
        () => {}
      );
      await maybePickGoogleAccount(popupPage);
    } else {
      await maybePickGoogleAccount(page);
    }

    await waitForMainApp(page);
    const dashboardScreenshot = await saveScreenshot(
      page,
      runFolder,
      "01-dashboard-loaded",
      true
    );
    report.Login = {
      status: "PASS",
      details: `Dashboard and sidebar visible. Screenshot: ${dashboardScreenshot}`
    };

    await openMiNegocioMenu(page);
    const menuScreenshot = await saveScreenshot(
      page,
      runFolder,
      "02-mi-negocio-menu-expanded",
      false
    );
    report["Mi Negocio menu"] = {
      status: "PASS",
      details: `Mi Negocio expanded with submenu options. Screenshot: ${menuScreenshot}`
    };

    const { cancelar } = await openAgregarNegocioModal(page);
    const modalScreenshot = await saveScreenshot(
      page,
      runFolder,
      "03-agregar-negocio-modal",
      false
    );
    report["Agregar Negocio modal"] = {
      status: "PASS",
      details: `Crear Nuevo Negocio modal validated. Screenshot: ${modalScreenshot}`
    };
    await clickAndWait(page, cancelar);

    await openAdministrarNegocios(page);
    const adminScreenshot = await saveScreenshot(
      page,
      runFolder,
      "04-administrar-negocios-view",
      true
    );
    report["Administrar Negocios view"] = {
      status: "PASS",
      details: `Account page sections validated. Screenshot: ${adminScreenshot}`
    };

    await validateInformacionGeneral(page);
    report["Información General"] = {
      status: "PASS",
      details: "User identity, plan text, and Cambiar Plan button validated."
    };

    await validateDetallesCuenta(page);
    report["Detalles de la Cuenta"] = {
      status: "PASS",
      details: "Cuenta creada, Estado activo, and Idioma seleccionado validated."
    };

    await validateTusNegocios(page);
    report["Tus Negocios"] = {
      status: "PASS",
      details: "Business list, Agregar Negocio button, and limit text validated."
    };

    const termsResult = await validateLegalLink({
      page,
      context,
      linkTextRegex: /términos y condiciones/i,
      headingRegex: /términos y condiciones/i,
      evidenceDir: runFolder,
      screenshotName: "05-terminos-y-condiciones"
    });
    report["Términos y Condiciones"] = {
      status: "PASS",
      details: `Legal content visible. URL: ${termsResult.finalUrl}. Screenshot: ${termsResult.screenshotPath}`
    };

    const privacyResult = await validateLegalLink({
      page,
      context,
      linkTextRegex: /política de privacidad/i,
      headingRegex: /política de privacidad/i,
      evidenceDir: runFolder,
      screenshotName: "06-politica-de-privacidad"
    });
    report["Política de Privacidad"] = {
      status: "PASS",
      details: `Legal content visible. URL: ${privacyResult.finalUrl}. Screenshot: ${privacyResult.screenshotPath}`
    };
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    for (const field of REPORT_FIELDS) {
      if (report[field].status === "PASS") {
        continue;
      }
      report[field] = {
        status: "FAIL",
        details: `Not completed because an earlier step failed. Root cause: ${message}`
      };
    }
  } finally {
    writeReportLine(report);
    const jsonReportPath = path.join(runFolder, "final_report.json");
    await fs.writeFile(jsonReportPath, JSON.stringify(report, null, 2), "utf-8");
    console.log(`Report JSON: ${jsonReportPath}`);
    await context.close();
    await browser.close();

    const allPassed = REPORT_FIELDS.every((field) => report[field].status === "PASS");
    if (!allPassed) {
      process.exitCode = 1;
    }
  }
}

run().catch((error) => {
  console.error("Unhandled error while running the SaleADS workflow test:", error);
  process.exitCode = 1;
});

File: src/test/java/e2e/SaleadsMiNegocioFullTest.java
package e2e;

import com.microsoft.playwright.Browser;
import com.microsoft.playwright.BrowserContext;
import com.microsoft.playwright.BrowserType;
import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.Playwright;
import com.microsoft.playwright.PlaywrightException;
import com.microsoft.playwright.options.LoadState;
import com.microsoft.playwright.options.WaitForSelectorState;
import org.junit.Assert;
import org.junit.Assume;
import org.junit.Test;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.LinkedHashMap;
import java.util.Map;

public class SaleadsMiNegocioFullTest {

  private static final String DEFAULT_GOOGLE_EMAIL = "juanlucasbarbiergarzon@gmail.com";
  private static final int DEFAULT_TIMEOUT_MS = 30000;
  private static final DateTimeFormatter FILE_TS = DateTimeFormatter.ofPattern("yyyyMMdd-HHmmss");

  @Test
  public void saleadsMiNegocioFullWorkflow() throws Exception {
    final boolean enabled = Boolean.parseBoolean(env("E2E_SALEADS_ENABLED", "false"));
    final String baseUrl = env("SALEADS_BASE_URL", "").trim();
    Assume.assumeTrue("Set E2E_SALEADS_ENABLED=true to run SaleADS E2E.", enabled);
    Assume.assumeTrue("Set SALEADS_BASE_URL to the active environment login page.", !baseUrl.isEmpty());

    final String googleEmail = env("SALEADS_GOOGLE_EMAIL", DEFAULT_GOOGLE_EMAIL);
    final boolean headless = Boolean.parseBoolean(env("E2E_HEADLESS", "true"));
    final Path evidenceDir = Paths.get(env("E2E_SCREENSHOT_DIR", "target/saleads-evidence"));
    Files.createDirectories(evidenceDir);

    final Map<String, StepStatus> report = initReport();
    final Map<String, String> legalUrls = new LinkedHashMap<>();

    try (Playwright playwright = Playwright.create()) {
      Browser browser = playwright.chromium().launch(new BrowserType.LaunchOptions().setHeadless(headless));
      BrowserContext context = browser.newContext(new Browser.NewContextOptions().setViewportSize(1440, 900));
      Page appPage = context.newPage();
      appPage.setDefaultTimeout(DEFAULT_TIMEOUT_MS);
      appPage.navigate(baseUrl);
      waitForUiSettle(appPage);

      runStep(report, "Login", () -> {
        Locator googleLoginButton = firstVisible(
          appPage,
          "button:has-text('Sign in with Google')",
          "button:has-text('Iniciar sesión con Google')",
          "button:has-text('Continuar con Google')",
          "button:has-text('Ingresar con Google')",
          "[role='button']:has-text('Google')"
        );
        clickAndWait(appPage, googleLoginButton);
        selectGoogleAccountIfPresent(appPage, context, googleEmail);

        requireAnyVisible(
          appPage,
          "Main application interface",
          "aside",
          "nav",
          "main",
          "text=/Mi\\s+Negocio/i"
        );
        requireAnyVisible(appPage, "Left sidebar navigation", "aside", "nav");
        screenshot(appPage, evidenceDir, "01-dashboard-loaded", false);
      });

      runStep(report, "Mi Negocio menu", () -> {
        Locator miNegocio = firstVisible(
          appPage,
          "text='Mi Negocio'",
          "text=/\\bNegocio\\b/i"
        );
        clickAndWait(appPage, miNegocio);

        requireAnyVisible(appPage, "'Agregar Negocio' visible", "text='Agregar Negocio'");
        requireAnyVisible(appPage, "'Administrar Negocios' visible", "text='Administrar Negocios'");
        screenshot(appPage, evidenceDir, "02-mi-negocio-menu-expanded", false);
      });

      runStep(report, "Agregar Negocio modal", () -> {
        Locator agregarNegocio = firstVisible(appPage, "text='Agregar Negocio'");
        clickAndWait(appPage, agregarNegocio);

        requireAnyVisible(appPage, "Modal title", "text='Crear Nuevo Negocio'");

        Locator businessNameInput = firstVisible(
          appPage,
          "input[placeholder='Nombre del Negocio']",
          "label:has-text('Nombre del Negocio') + input",
          "input[name*='negocio']"
        );
        Assert.assertTrue("Nombre del Negocio input should be visible", businessNameInput.isVisible());

        requireAnyVisible(appPage, "Limit text", "text='Tienes 2 de 3 negocios'");
        requireAnyVisible(appPage, "Cancelar button", "button:has-text('Cancelar')");
        requireAnyVisible(appPage, "Crear Negocio button", "button:has-text('Crear Negocio')");

        screenshot(appPage, evidenceDir, "03-agregar-negocio-modal", false);

        businessNameInput.fill("Negocio Prueba Automatización");
        clickAndWait(appPage, firstVisible(appPage, "button:has-text('Cancelar')"));
      });

      runStep(report, "Administrar Negocios view", () -> {
        if (!isAnyVisible(appPage, "text='Administrar Negocios'")) {
          clickAndWait(appPage, firstVisible(appPage, "text='Mi Negocio'", "text=/\\bNegocio\\b/i"));
        }
        clickAndWait(appPage, firstVisible(appPage, "text='Administrar Negocios'"));

        requireAnyVisible(appPage, "Información General section", "text='Información General'");
        requireAnyVisible(appPage, "Detalles de la Cuenta section", "text='Detalles de la Cuenta'");
        requireAnyVisible(appPage, "Tus Negocios section", "text='Tus Negocios'");
        requireAnyVisible(appPage, "Sección Legal section", "text='Sección Legal'");

        screenshot(appPage, evidenceDir, "04-administrar-negocios-page", true);
      });

      runStep(report, "Información General", () -> {
        requireAnyVisible(appPage, "User name visible", "[data-testid*='name']", "text=/^[A-Z][a-z]+\\s+[A-Z].*/");
        requireAnyVisible(appPage, "User email visible", "text=/[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}/");
        requireAnyVisible(appPage, "BUSINESS PLAN text", "text='BUSINESS PLAN'");
        requireAnyVisible(appPage, "Cambiar Plan button", "button:has-text('Cambiar Plan')");
      });

      runStep(report, "Detalles de la Cuenta", () -> {
        requireAnyVisible(appPage, "Cuenta creada", "text='Cuenta creada'");
        requireAnyVisible(appPage, "Estado activo", "text='Estado activo'");
        requireAnyVisible(appPage, "Idioma seleccionado", "text='Idioma seleccionado'");
      });

      runStep(report, "Tus Negocios", () -> {
        requireAnyVisible(
          appPage,
          "Business list visible",
          "text='Tus Negocios'",
          "table",
          "[role='table']",
          "[role='list']"
        );
        requireAnyVisible(appPage, "Agregar Negocio button exists", "button:has-text('Agregar Negocio')", "text='Agregar Negocio'");
        requireAnyVisible(appPage, "Tienes 2 de 3 negocios text", "text='Tienes 2 de 3 negocios'");
      });

      runStep(report, "Términos y Condiciones", () -> {
        String url = validateLegalLinkAndReturn(
          appPage,
          "Términos y Condiciones",
          "Términos y Condiciones",
          evidenceDir,
          "08-terminos-y-condiciones"
        );
        legalUrls.put("Términos y Condiciones", url);
      });

      runStep(report, "Política de Privacidad", () -> {
        String url = validateLegalLinkAndReturn(
          appPage,
          "Política de Privacidad",
          "Política de Privacidad",
          evidenceDir,
          "09-politica-de-privacidad"
        );
        legalUrls.put("Política de Privacidad", url);
      });

      writeFinalReport(report, legalUrls, evidenceDir);
      context.close();
      browser.close();
    }

    Assert.assertTrue("Final report includes failed validations:\n" + summarize(report), allPassed(report));
  }

  private static String validateLegalLinkAndReturn(Page appPage,
                                                   String linkText,
                                                   String headingText,
                                                   Path evidenceDir,
                                                   String screenshotName) {
    String appUrlBefore = appPage.url();
    Page targetPage;
    boolean popupOpened = false;

    try {
      targetPage = appPage.waitForPopup(
        new Page.WaitForPopupOptions().setTimeout(5000),
        () -> clickAndWait(appPage, firstVisible(appPage, "text='" + linkText + "'"))
      );
      popupOpened = true;
    } catch (PlaywrightException ex) {
      clickAndWait(appPage, firstVisible(appPage, "text='" + linkText + "'"));
      targetPage = appPage;
    }

    waitForUiSettle(targetPage);
    requireAnyVisible(targetPage, headingText + " heading", "text='" + headingText + "'");
    requireAnyVisible(
      targetPage,
      headingText + " legal text",
      "article",
      "main",
      "text=/t[eé]rminos|privacidad|legal|condiciones/i"
    );
    screenshot(targetPage, evidenceDir, screenshotName, true);

    String finalUrl = targetPage.url();

    if (popupOpened) {
      targetPage.close();
      appPage.bringToFront();
      waitForUiSettle(appPage);
    } else {
      appPage.goBack();
      if (!appPage.url().equals(appUrlBefore)) {
        appPage.navigate(appUrlBefore);
      }
      waitForUiSettle(appPage);
    }

    return finalUrl;
  }

  private static void runStep(Map<String, StepStatus> report, String stepName, Runnable action) {
    try {
      action.run();
      report.put(stepName, StepStatus.pass());
    } catch (Throwable t) {
      report.put(stepName, StepStatus.fail(t.getMessage()));
    }
  }

  private static void clickAndWait(Page page, Locator locator) {
    locator.first().waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.VISIBLE));
    locator.first().click();
    waitForUiSettle(page);
  }

  private static void waitForUiSettle(Page page) {
    try {
      page.waitForLoadState(LoadState.DOMCONTENTLOADED);
      page.waitForLoadState(LoadState.NETWORKIDLE, new Page.WaitForLoadStateOptions().setTimeout(5000));
    } catch (PlaywrightException ignored) {
      // Not every click triggers navigation/network-idle; keep progressing.
    }
    page.waitForTimeout(500);
  }

  private static void selectGoogleAccountIfPresent(Page appPage, BrowserContext context, String email) {
    Locator accountOption = null;

    if (isAnyVisible(appPage, "text='" + email + "'")) {
      accountOption = appPage.locator("text='" + email + "'");
    } else {
      for (Page candidate : context.pages()) {
        if (candidate.isClosed()) {
          continue;
        }
        if (isAnyVisible(candidate, "text='" + email + "'")) {
          accountOption = candidate.locator("text='" + email + "'");
          break;
        }
      }
    }

    if (accountOption != null && accountOption.first().isVisible()) {
      accountOption.first().click();
      waitForUiSettle(appPage);
    }
  }

  private static Locator firstVisible(Page page, String... selectors) {
    long deadline = System.currentTimeMillis() + DEFAULT_TIMEOUT_MS;
    while (System.currentTimeMillis() < deadline) {
      for (String selector : selectors) {
        Locator candidate = page.locator(selector).first();
        try {
          if (candidate.count() > 0 && candidate.isVisible()) {
            return candidate;
          }
        } catch (PlaywrightException ignored) {
          // Keep searching across selectors until timeout.
        }
      }
      page.waitForTimeout(250);
    }
    throw new AssertionError("No visible selector found: " + String.join(" | ", selectors));
  }

  private static void requireAnyVisible(Page page, String description, String... selectors) {
    Locator visible = firstVisible(page, selectors);
    Assert.assertTrue(description + " should be visible", visible.isVisible());
  }

  private static boolean isAnyVisible(Page page, String... selectors) {
    for (String selector : selectors) {
      Locator candidate = page.locator(selector).first();
      try {
        if (candidate.count() > 0 && candidate.isVisible()) {
          return true;
        }
      } catch (PlaywrightException ignored) {
        // Ignore transient selector failures and continue.
      }
    }
    return false;
  }

  private static void screenshot(Page page, Path evidenceDir, String checkpoint, boolean fullPage) {
    String fileName = LocalDateTime.now().format(FILE_TS) + "-" + checkpoint + ".png";
    Path path = evidenceDir.resolve(fileName);
    page.screenshot(new Page.ScreenshotOptions().setPath(path).setFullPage(fullPage));
  }

  private static Map<String, StepStatus> initReport() {
    Map<String, StepStatus> report = new LinkedHashMap<>();
    report.put("Login", StepStatus.notRun());
    report.put("Mi Negocio menu", StepStatus.notRun());
    report.put("Agregar Negocio modal", StepStatus.notRun());
    report.put("Administrar Negocios view", StepStatus.notRun());
    report.put("Información General", StepStatus.notRun());
    report.put("Detalles de la Cuenta", StepStatus.notRun());
    report.put("Tus Negocios", StepStatus.notRun());
    report.put("Términos y Condiciones", StepStatus.notRun());
    report.put("Política de Privacidad", StepStatus.notRun());
    return report;
  }

  private static void writeFinalReport(Map<String, StepStatus> report, Map<String, String> legalUrls, Path evidenceDir) throws Exception {
    StringBuilder sb = new StringBuilder();
    sb.append("SaleADS Mi Negocio full workflow report\n");
    sb.append("Generated at: ").append(LocalDateTime.now()).append("\n\n");
    sb.append("Validation status:\n");
    for (Map.Entry<String, StepStatus> entry : report.entrySet()) {
      sb.append("- ").append(entry.getKey()).append(": ").append(entry.getValue().status);
      if (!entry.getValue().reason.isEmpty()) {
        sb.append(" (").append(entry.getValue().reason).append(")");
      }
      sb.append("\n");
    }

    sb.append("\nFinal URLs:\n");
    sb.append("- Términos y Condiciones: ").append(legalUrls.getOrDefault("Términos y Condiciones", "N/A")).append("\n");
    sb.append("- Política de Privacidad: ").append(legalUrls.getOrDefault("Política de Privacidad", "N/A")).append("\n");

    Path reportFile = evidenceDir.resolve("final-report.txt");
    Files.writeString(reportFile, sb.toString());
    System.out.println(sb);
  }

  private static boolean allPassed(Map<String, StepStatus> report) {
    return report.values().stream().allMatch(status -> "PASS".equals(status.status));
  }

  private static String summarize(Map<String, StepStatus> report) {
    StringBuilder sb = new StringBuilder();
    for (Map.Entry<String, StepStatus> entry : report.entrySet()) {
      sb.append(entry.getKey()).append(": ").append(entry.getValue().status);
      if (!entry.getValue().reason.isEmpty()) {
        sb.append(" (").append(entry.getValue().reason).append(")");
      }
      sb.append("\n");
    }
    return sb.toString();
  }

  private static String env(String key, String defaultValue) {
    String value = System.getenv(key);
    return value == null || value.isBlank() ? defaultValue : value;
  }

  private static final class StepStatus {
    private final String status;
    private final String reason;

    private StepStatus(String status, String reason) {
      this.status = status;
      this.reason = reason == null ? "" : reason;
    }

    static StepStatus pass() {
      return new StepStatus("PASS", "");
    }

    static StepStatus fail(String reason) {
      return new StepStatus("FAIL", reason);
    }

    static StepStatus notRun() {
      return new StepStatus("NOT_RUN", "");
    }
  }
}
